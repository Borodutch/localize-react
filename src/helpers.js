export const WARNING_MESSAGE = 'localize-react: There are no translations for specified locale';

export function sanitizeLocale(locale, translations) {
  if (!locale) {
    return null;
  }

  if (typeof translations[locale] === 'object') {
    return locale;
  }

  const normalizedLocale = locale.toLowerCase().replace(/-/g, '_');
  if (typeof translations[normalizedLocale] === 'object') {
    return normalizedLocale;
  }

  const shortenLocale = normalizedLocale.split('_')[0];
  if (typeof translations[shortenLocale] === 'object') {
    return shortenLocale;
  }

  console.warn(WARNING_MESSAGE, locale);
  return locale;
}
