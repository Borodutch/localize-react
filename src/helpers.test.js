import { clearCache, memoize, sanitizeLocale, WARNING_MESSAGE } from './helpers';

describe('helpers', () => {
  describe('sanitizeLocale', () => {
    it('returns null is locale is not specified', () => {
      const sanitizedLocale = sanitizeLocale(null, {});
      expect(sanitizedLocale).toMatchSnapshot();
    });

    it('returns the same locale is that is defined in translations', () => {
      const sanitizedLocale = sanitizeLocale('En-uS', { 'En-uS': { name: 'Alex' } });
      expect(sanitizedLocale).toMatchSnapshot();
    });

    it('returns the lowercase underscore locale if original was not found', () => {
      const sanitizedLocale = sanitizeLocale('En-cA', { en_ca: { name: 'Alex' } });
      expect(sanitizedLocale).toMatchSnapshot();
    });

    it('returns the shorten locale if original was not found', () => {
      const sanitizedLocale = sanitizeLocale('En-Fr', { en: { name: 'Alex' } });
      expect(sanitizedLocale).toMatchSnapshot();
    });

    it('returns the original and makes warning if no matching happened', () => {
      global.console = { warn: jest.fn() };

      const sanitizedLocale = sanitizeLocale('FR-CA', { en: { name: 'Alex' } });
      expect(sanitizedLocale).toMatchSnapshot();
      expect(global.console.warn).toHaveBeenCalled();
      expect(global.console.warn).toHaveBeenCalledWith(WARNING_MESSAGE, 'FR-CA');
    });
  });

  describe('memoize', () => {
    it('memoizes value', () => {
      const fn = jest.fn(_ => _);

      const memoizedFn = memoize(fn);
      const firstResult = memoizedFn(1);
      const secondResult = memoizedFn(1);

      clearCache();

      expect(firstResult).toEqual(secondResult);
      expect(fn).toHaveBeenCalledTimes(1);
    });
  });
});

