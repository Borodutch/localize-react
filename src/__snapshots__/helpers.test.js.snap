// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers buildTranslation builds simple translation with empty values 1`] = `"Hello {{world}}"`;

exports[`helpers buildTranslation builds simple translation with no values 1`] = `"Hello {{world}}"`;

exports[`helpers buildTranslation builds translation with multiple values 1`] = `"Hello, my name is Alex. I am 25"`;

exports[`helpers buildTranslation builds translation with no templates inside 1`] = `"Hello"`;

exports[`helpers sanitizeLocale returns null is locale is not specified 1`] = `null`;

exports[`helpers sanitizeLocale returns the lowercase underscore locale if original was not found 1`] = `"en_ca"`;

exports[`helpers sanitizeLocale returns the original and makes warning if no matching happened 1`] = `"FR-CA"`;

exports[`helpers sanitizeLocale returns the same locale is that is defined in translations 1`] = `"En-uS"`;

exports[`helpers sanitizeLocale returns the shorten locale if original was not found 1`] = `"en"`;

exports[`helpers transformToPairs creates pair of template - value 1`] = `
Array [
  Array [
    "{{name}}",
    "Alex",
  ],
]
`;

exports[`helpers transformToPairs creates pairs of template - value 1`] = `
Array [
  Array [
    "{{name}}",
    "Alex",
  ],
  Array [
    "{{age}}",
    25,
  ],
]
`;

exports[`helpers transformToPairs creates pairs with same value if no correspondent value found 1`] = `
Array [
  Array [
    "{{name}}",
    "Alex",
  ],
  Array [
    "{{age}}",
    25,
  ],
  Array [
    "{{not_found}}",
    "{{not_found}}",
  ],
]
`;
